apache2/apache2_util.c

```diff
@@ -187,6 +192,115 @@ char *get_env_var(request_rec *r, char *name) {
     return result;
 }
 
+#ifdef WAF_JSON_LOGGING_ENABLE
+/**
+ * Retrieve waf log field.
+ */
+void get_field_value(const char* from, const char* to, const char* text, char* output) {
+    char* first = strstr(text, from);
+    int first_index = first - text;
+
+    if (first != NULL ) {
+        if ((first_index > 0 && (first-1)[0] != "\\") || (first_index == 0)) {
+            first += strlen(from);
+        }
+        else {
+            first = NULL;
+        }
+    }
+
+    char* last = strstr(first, to);
+    int last_index = last- first;
+    if (last != NULL ) {
+        if ((last_index > 0 && (last-1)[0] != "\\") || (last_index == 0)) {
+        }
+        else {
+            last = NULL;
+        }
+    }
+
+    if (first != NULL && last != NULL) {
+        strncpy(output, first, last - first);
+    }
+}
+
+/**
+ * get ip and port number.
+ */
+void get_ip_port(const char* ip_port, char* waf_ip, char* waf_port) {
+    char *comma = strstr(ip_port, ":");
+    if (comma!= NULL) {
+        strcpy(waf_port, comma+1);
+        strncpy(waf_ip, ip_port, comma - ip_port);
+    }
+    else {
+        strcpy(waf_ip, ip_port);
+    }
+}
+
+/**
+ * get detail_messages.
+ */
+void get_detail_message(const char* str1, char* waf_detail_message) {
+    char *end = strstr(str1, "[file ");
+    if (end != NULL) {
+        strncpy(waf_detail_message, str1, end - str1);
+    }
+}
+
+/**
+ * only expose short path.
+ */
+void get_short_filename(char* waf_filename) {
+    char tmp_filename[1024] = ""; 
+    char *index = strstr(waf_filename, WAF_RULESET_PREFIX);
+
+    if (index != NULL) {
+        index += strlen(WAF_RULESET_PREFIX);
+        index = strstr(index, "/");
+        if (index != NULL) {
+            strcpy(tmp_filename, index+1);
+            strcpy(waf_filename, tmp_filename);
+        }
+    }
+}
+
+/**
+ * send all waf fields in json format to a file.
+ */
+void send_waf_log(const char* data_dir, const char* str1, const char* ip_port, const char* uri, const char* time, int mode, const char* hostname, request_rec *r) {
+    int rc = 0;
+    char waf_filename[1024] = "";
```

Comment:
what happens if filename is > 1024?

Reply:
the maxium of str1 is 1024, so won't exceed this value